### PR TITLE
[libosmscout] Add new port

### PR DESCRIPTION
**Pull request by OgreTransporter**
This is a new port of libosmscout (https://github.com/Framstag/libosmscout).

### CODE REVIEW

**Comment by PhoebeHui** [comment]
@OgreTransporter, could your format the vcpkg.json file? See github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#manifest for solution.

File: ports/libosmscout/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		directx OSMDCOUT_BUILD_MAP_DIRECTX
+		gdi OSMDCOUT_BUILD_MAP_GDI
+		svg OSMDCOUT_BUILD_MAP_SVG
+		qt OSMDCOUT_BUILD_MAP_QT
+	)
+else()
+	vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+		cairo OSMDCOUT_BUILD_MAP_CAIRO
+		svg OSMDCOUT_BUILD_MAP_SVG
+		qt OSMDCOUT_BUILD_MAP_QT
```
Comment (PhoebeHui) [inline ports/libosmscout/portfile.cmake:21]: Have you test these features locally?

File: ports/libosmscout/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        -DOSMSCOUT_BUILD_EXTERN_MATLAB=OFF
+        -DOSMSCOUT_BUILD_TESTS=OFF
+		${FEATURE_OPTIONS}
+)
+
+vcpkg_install_cmake()
+vcpkg_copy_tools(TOOL_NAMES Import AUTO_CLEAN)
+vcpkg_fixup_cmake_targets(CONFIG_PATH share/cmake/libosmscout)
+
+file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/libosmscout RENAME copyright)
```
Comment (PhoebeHui) [inline ports/libosmscout/portfile.cmake:49]:
```suggestion
file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)

```

**Review by PhoebeHui — CHANGES_REQUESTED** [review]

File: ports/libosmscout/portfile.cmake (hunk 1 above)
Comment (OgreTransporter) [reply ports/libosmscout/portfile.cmake:21]: Yes, I've tested it here with Visual Studio 2017.

**Comment by OgreTransporter** [comment]
> @OgreTransporter, could your format the vcpkg.json file? See https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#manifests for solution.

I would be happy about a reasonable error message. 
> The formatting of the manifest files didn't match our expectation.

is not very helpful. Of course I used `vcpkg.exe format-manifest --all` before. What else is wrong?

> I also suggest to rename the port to 'framstag-libosmscout', I noticed there is also another package named to libosmscout. https://repology.org/project/libosmscout/versions

I am against renaming the port. If the name is changed, repology probably cannot recognize the vcpkg package. For example, cairo is also called cairo everywhere. The listed packages have also not been maintained for years and are very outdated.


```
C:\dev\vcpkg>vcpkg.exe format-manifest --all
Succeeded in formatting the manifest files.

C:\dev\vcpkg>vcpkg.exe install libosmscout[cairo,directx,gdi,svg]:x64-windows
Computing installation plan...
The following packages will be built and installed:
    libosmscout[cairo,core,directx,gdi,svg]:x64-windows
Detecting compiler hash for triplet x64-windows...
Starting package 1/1: libosmscout:x64-windows
Building package libosmscout[cairo,core,directx,gdi,svg]:x64-windows...
Could not locate cached archive: C:\Users\IEUser\AppData\Local\vcpkg\archives\97\97a62f29ccc6aa6ac88f947eaada27885c0c6b48.zip
-- Using cached C:/dev/vcpkg/downloads/Framstag-libosmscout-c81e1d9a0f69cc5b93588dbe330b2af587162c5f.tar.gz
-- Cleaning sources at C:/dev/vcpkg/buildtrees/libosmscout/src/f587162c5f-4dbef76e40.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/dev/vcpkg/downloads/Framstag-libosmscout-c81e1d9a0f69cc5b93588dbe330b2af587162c5f.tar.gz
-- Using source at C:/dev/vcpkg/buildtrees/libosmscout/src/f587162c5f-4dbef76e40.clean
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: C:/dev/vcpkg/packages/libosmscout_x64-windows/share/libosmscout/copyright
-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: C:\Users\IEUser\AppData\Local\vcpkg\archives\97\97a62f29ccc6aa6ac88f947eaada27885c0c6b48.zip
Building package libosmscout[cairo,core,directx,gdi,svg]:x64-windows... done
Installing package libosmscout[cairo,core,directx,gdi,svg]:x64-windows...
Installing package libosmscout[cairo,core,directx,gdi,svg]:x64-windows... done
Elapsed time for package libosmscout:x64-windows: 19.09 min

Total elapsed time: 19.17 min

The package libosmscout:x64-windows provides CMake targets:

    find_package(libosmscout CONFIG REQUIRED)
    target_link_libraries(main PRIVATE OSMScout::OSMScout)


C:\dev\vcpkg>
```

**Comment by PhoebeHui** [comment]
For the name issue, @strega-nil, could you help confirm if it's okay in vcpkg?

Here is CI test failures, @OgreTransporter, could you confirm if it supports uwp and arm? let me know if you need full logs.
arm64-windows/arm-uwp:
```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.27.29110\include\xmmintrin.h(31,1): fatal error C1189: #error:  This header is specific to X86 and X64 targets 
```

x64-uwp:
```
C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(486,5): error MSB4181: The "CompileXaml" task returned false but did not log an error. [D:\buildtrees\libosmscout\x64-uwp-dbg\libosmscout\OSMScout.vcxproj]

    D:\buildtrees\libosmscout\src\f587162c5f-4dbef76e40.clean\libosmscout\src\osmscout\util\File.cpp(157,46): error C2664: 'DWORD GetFileAttributesW(LPCWSTR)': cannot convert argument 1 from 'const _Elem *' to 'LPCWSTR' [D:\buildtrees\libosmscout\x64-uwp-dbg\libosmscout\OSMScout.vcxproj]
         D:\buildtrees\libosmscout\src\f587162c5f-4dbef76e40.clean\libosmscout\src\osmscout\util\File.cpp(169,56): error C2664: 'DWORD GetFileAttributesW(LPCWSTR)': cannot convert argument 1 from 'const _Elem *' to 'LPCWSTR' [D:\buildtrees\libosmscout\x64-uwp-dbg\libosmscout\OSMScout.vcxproj]
```

x64-windows-static:
```
CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScout>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.

CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutImport>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.

CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutGPX>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.

CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutMap>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.

CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutMapDirectX>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.
CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutMapGDI>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.
CMake Error:
  Error evaluating generator expression:
    $<TARGET_PDB_FILE:OSMScoutMapSVG>
  TARGET_PDB_FILE is allowed only for targets with linker created artifacts.
```

**Comment by OgreTransporter** [comment]
@PhoebeHui Thanks a lot for your help! I find it strange that the Linux build fails because the CI (https://ci.appveyor.com/project/Framstag/libosmscout/builds/36178606) of libosmscout works. What kind of Linux and which settings were used? static-windows I have fixed, UWP and ARM I will probably disable.

**Comment by OgreTransporter** [comment]
I can build libsomscout statically with CMake and my own libraries, but not with vcpkg:
```
Building package glib[core]:x64-windows-static...
Could not locate cached archive: C:\Users\IEUser\AppData\Local\vcpkg\archives\b0\b0d4ff0d5c509e12b835275d1b95cfd508ff88b1.zip
-- Note: glib only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:42 (message):
  Refusing to build unexpected dynamic library against the static CRT.  If
  this is desired, please configure your triplet to directly request this
  configuration.
Call Stack (most recent call first):
  ports/glib/portfile.cmake:6 (vcpkg_check_linkage)
  scripts/ports.cmake:135 (include)


Error: Building package glib:x64-windows-static failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: glib:x64-windows-static
  Vcpkg version: 2020.06.15-nohash

Additionally, attach any relevant sections from the log files above.
```
If the bug #12094 is fixed, I activate it again.

**Comment by PhoebeHui** [comment]
@OgreTransporter, for linux, the CI test machine using ubuntu 18.04 with gnu 7.5.0, this issue could repro locally in my vm. I attached the full log:
[failure logs for x64-linux (1).zip](https://github.com/microsoft/vcpkg/files/5522616/failure.logs.for.x64-linux.1.zip)

For CI test machine deployment in linux, see https://github.com/microsoft/vcpkg/blob/master/scripts/azure-pipelines/linux/provision-image.sh

Noted that the linux passed in CI testing now since you add '!static' in supports, it actually skipped.

**Comment by PhoebeHui** [comment]
All features test passed with x64-windows.

**Comment by OgreTransporter** [comment]
When I found the Linux error I will activate it again. The CIs of libosmscout work under Ubuntu 18.04:
 - https://ci.appveyor.com/project/Framstag/libosmscout/builds/36178606
 - https://github.com/Framstag/libosmscout/runs/1365208594

Windows static does not work for me because glib (dependency of a dependency) cannot be built. So, the package is only available as Windows DLLs for now. Later I can update it.

**Comment by OgreTransporter** [comment]
I still do not understand the renaming. The ports of many libraries are named the same as the packages on different systems, like [boost](https://repology.org/project/boost/versions), [bzip2](https://repology.org/project/bzip2/versions), [cairo](https://repology.org/project/cairo/versions), [cgal](https://repology.org/project/cgal/versions), [curl](https://repology.org/project/curl/versions), etc. Isn't it also the sense that you can find a known library by its name again? But why should libosmscout be renamed?

**Comment by PhoebeHui** [comment]
There is 2 packages named to libosmscout, and users may confused with the names, see [doc](https://github.com/microsoft/vcpkg/blob/0fec1340eba828e95c489796eb0c7a4330120686/docs/maintainers/manifest-files.md#name)

https://github.com/Framstag/libosmscout
https://github.com/OpenMandrivaAssociation/libosmscout

**Comment by OgreTransporter** [comment]
Both are the same library, except that the Open Mandriva Association directory is outdated and https://github.com/Framstag/libosmscout is the official repository. The Mandriva directory is just an outdated build script like a vcpkg port.

**Comment by strega-nil** [comment]
Confirmed that libosmscout is the correct name.

**Comment by OgreTransporter** [comment]
Many thanks for your help.

**Comment by strega-nil** [comment]
Thank you @OgreTransporter ! :)

**Pull request merged**